In [2]:
import pandas as pd
import numpy as np

In [3]:
pd.options.display.max_columns = None

# Pair Programming EDA I
Empezamos una nueva lección, y este es el momento de poner en práctica los conocimientos adquiridos del Análisis Exploratorio de Datos. 

Si recordamos, en la lecciones de Pandas estuvimos usando un dataset sobre ataques de tiburones, y que hicimos algunos cambios a lo largo de los ejercicios, incluimos filas, columnas e incluso limpiamos algunas colummas 💪🏽! Para todas los ejercicios de pair programming de EDA I y II usaremos el dataframe generado en el ejercicio de Pandas VII.

En el ejercicio de hoy nos familiarizaremos con el dataset completo sobre el que trabajaremos en la parte de limpieza.


Pero para hacer una exploración buena del dataset no lo podemos hacer como pollo sin cabeza a ver que nos encontramos. Necesitamos hacernos una serie de preguntas o hipótesis que queramos contestar. En este caso os plantemos las siguientes preguntas a contestar a lo largo de los ejercicios de pair programming de este módulo:

Donde variable es cada una de las variables categóricas que tenemos en el dataset y n_unicos es una lista con los nombres de los valores únicos de cada variable.

In [4]:
df = pd.read_csv("../data/00-tiburon_4.csv",index_col=0)

# Hipótesis

Las hipótesis que os planteamos son:

- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los países en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?


# Objetivos
Los objetivos de hoy son:
Haced una exploración inicial del dataframe:

- ¿Cuál es el número de filas y columnas?
- ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.
- ¿Qué tipos de datos hay en cada columna?
- Extraed la información general del dataframe.
- Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.
- ¿Cuáles son los principales estádisticos de los dataframes?
- Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe.

In [5]:
df.head()

case_number    year        type    country        area  \
0    2018.06.25  2018.0     Boating        USA  California   
1  2018.06.03.a  2018.0  Unprovoked     BRAZIL  Pernambuco   
2  2018.05.26.b  2018.0  Unprovoked        USA     Florida   
3    2018.05.24  2018.0    Provoked  AUSTRALIA  Queensland   
4  2018.05.13.a  2018.0     Invalid    ENGLAND    Cornwall   

                       location        activity                    name age  \
0   Oceanside, San Diego County        Paddling             Julie Wolfe  57   
1         Piedade Beach, Recife        Swimming  Jose Ernesto da Silva   18   
2  Cocoa Beach, Brevard  County         Walking               Cody High  15   
3               Cairns Aquarium  Feeding sharks                    male  32   
4                Off Land's End         Fishing            Max Berryman  21   

             time          species_  \
0           18h00       White shark   
1  Late afternoon       Tiger shark   
2           17h00    Bull shark, 6'   
3         Morning   Grey reef shark   
4           08h15  Invalid incident   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury  mes deceso genero  
0  No injury to occupant, outrigger canoe and pad...  Jun      N      F  
1                                              FATAL  Jun      Y      M  
2                              Lower left leg bitten  May      N      M  
3  Minor bite to hand by captive shark. PROVOKED ...  May      N      M  
4  Injured by teeth of a dead porbeagle shark he ...  May      N      M

- ¿Cuál es el número de filas y columnas?

In [6]:
df.shape

(1672, 17)

#### Tiene 1672 filas y 17 columnas

- ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [7]:
df_nulos = pd.DataFrame(df.isnull().sum()*100/df.shape[0])
df_nulos

0
case_number  0.000000
year         0.000000
type         0.000000
country      0.598086
area         2.751196
location     3.050239
activity     1.854067
name         0.538278
age          9.210526
time         9.569378
species_     7.535885
href         0.000000
siglo        0.000000
injury       0.478469
mes          9.210526
deceso       5.921053
genero       0.837321

- ¿Qué tipos de datos hay en cada columna?

In [8]:
df.dtypes

case_number     object
year           float64
type            object
country         object
area            object
location        object
activity        object
name            object
age             object
time            object
species_        object
href            object
siglo           object
injury          object
mes             object
deceso          object
genero          object
dtype: object

- Extraed la información general del dataframe.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  injury       1664 non-null   object 
 14  mes          1518 non-null   object 
 15  deceso       1573 non-null   object 
 16  genero       1658 non-null   object 
dtypes: float64(1), object(16)
memory usage: 235.1+ KB


- Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [10]:
# Creamos un dataframe nuevo con las variables categóricas
df_categoricas = df.select_dtypes(include = "object")
df_categoricas.head(2)

case_number        type country        area                     location  \
0    2018.06.25     Boating     USA  California  Oceanside, San Diego County   
1  2018.06.03.a  Unprovoked  BRAZIL  Pernambuco        Piedade Beach, Recife   

   activity                    name age            time     species_  \
0  Paddling             Julie Wolfe  57           18h00  White shark   
1  Swimming  Jose Ernesto da Silva   18  Late afternoon  Tiger shark   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury  mes deceso genero  
0  No injury to occupant, outrigger canoe and pad...  Jun      N      F  
1                                              FATAL  Jun      Y      M

In [11]:
# Creamos un dataframe nuevo con las variables numéricas
df_numericas = df.select_dtypes(include="float64")
df_numericas.head(2)

year
0  2018.0
1  2018.0

- ¿Cuáles son los principales estádisticos de los dataframes?

In [12]:
df_numericas.describe().T

count         mean         std  min     25%     50%     75%     max
year  1672.0  1832.688397  528.969901  0.0  1964.0  1997.0  2008.0  2018.0

In [13]:
df_categoricas.describe(include="object").T

count unique                                                top  \
case_number  1672   1672                                         2018.06.25   
type         1672      6                                         Unprovoked   
country      1662     95                                                USA   
area         1626    236                                            Florida   
location     1621   1310                   New Smyrna Beach, Volusia County   
activity     1641    404                                            Surfing   
name         1663   1536                                               male   
age          1518     94                                                 19   
time         1512    248                                              11h00   
species_     1546    856                                        White shark   
href         1672   1672  http://sharkattackfile.net/spreadsheets/pdf_di...   
siglo        1672      2                                           siglo-XX   
injury       1664   1173                                              FATAL   
mes          1518     12                                                Aug   
deceso       1573      2                                                  N   
genero       1658      2                                                  M   

             freq  
case_number     1  
type         1432  
country       741  
area          346  
location       73  
activity      426  
name           97  
age            70  
time           74  
species_       76  
href            1  
siglo         887  
injury        154  
mes           162  
deceso       1256  
genero       1435

- Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe.

In [14]:
# Creamos un Dataframe a partir de un diccionario.
# Creamos dos listas para meter el nombre de la variable y los valores unicos
variable = []
n_uniques = []
for col in df_categoricas.columns: # Iteramos por las columnas de DF categoricas y apendeamos los distintos valores
    variable.append(col)
    n_uniques.append(df_categoricas[col].unique())
datos = {"variable": variable, "n_uniques": n_uniques} # Creamos el diccionario

In [15]:
# Creamos el Dataframe a partir del diccionario
df_valoresunicos = pd.DataFrame.from_dict(datos)

In [16]:
df_valoresunicos

variable                                          n_uniques
0   case_number  [2018.06.25, 2018.06.03.a, 2018.05.26.b, 2018....
1          type  [Boating, Unprovoked, Provoked, Invalid, Sea D...
2       country  [USA, BRAZIL, AUSTRALIA, ENGLAND, COSTA RICA, ...
3          area  [California, Pernambuco, Florida, Queensland, ...
4      location  [Oceanside, San Diego County, Piedade Beach, R...
5      activity  [Paddling, Swimming, Walking, Feeding sharks, ...
6          name  [Julie Wolfe, Jose Ernesto da Silva , Cody Hig...
7           age  [57, 18, 15, 32, 21, 30, 60, 33, 19, 25, 10, 6...
8          time  [18h00, Late afternoon, 17h00, Morning, 08h15,...
9      species_  [White shark, Tiger shark, Bull shark, 6', Gre...
10         href  [http://sharkattackfile.net/spreadsheets/pdf_d...
11        siglo                              [siglo-XXI, siglo-XX]
12       injury  [No injury to occupant, outrigger canoe and pa...
13          mes  [Jun, May, Apr, Mar, Feb, Jan, Dec, Nov, Oct, ...
14       deceso                                        [N, Y, nan]
15       genero                                        [F, M, nan]

In [17]:
df_valoresunicos.loc[8, "n_uniques"]

array(['18h00', 'Late afternoon', '17h00', 'Morning', '08h15', '10h40',
       '09h30', '21h50', '09h40', '15h00', '17h35', '15h30', '07h30',
       '19h00, Dusk', '16h00', '13h45', '23h30', '09h00', 'Afternoon',
       '18h30', '12h30', '18h45', '06h00', '10h30', 'Midday', '10h00',
       '13h30', '11h20', '07h45', '08h30', 'Shortly before 12h00',
       '14h30', '12h00', '07h00', '02h00', '11h00', '9h00', '15h15',
       '16h30', '11h30', 'After noon', '22h00', '15h25', '17h30', '15h45',
       'Morning ', '13h20', '10h45', '10h15', '18h15', '14h00', '13h50',
       '19h20', '10h25', '16h20', '16h45', '15h52', '06h15', '14h40',
       '08h00', '19h00', '10h10', '08h45', '12h10', '18h05', '12h25',
       '17h51', '16h12', '12h45', '10h50', '13h00', '09h45', '11h45',
       '06h45', '05h00', '03h30', '06h30', '20h00', '16h40', '17h50',
       'Night', '18h15-18h30', '08h20', '20h30', '15h19', '15h55',
       '16h05', '14h10', '12h50', '11h15', '13h40', '13h24',
       '09h00 - 09h30', 

# Conclusiones
De qué nos sirven estos números y dataframe? Extraed conclusiones de ellos.

Por ejemplo:

- ¿Hay alguna columna que no tenga sentido su tipo de datos?
> El año no tiene sentido que sea un float, tendría que ser datetime o integer.

> La edad tampoco tiene sentido que sea un object, debería ser un integer

> La columna time en realidad tiene una mezcla de datos, que algunos deberían ser datime y otros object
- ¿Hay alguna columna que no tenga sentido su nombre?
> Podríamos quitar la barra baja de species_

> La columna href tampoco tiene mucho sentido. Podríamos llamarla mejor pdf_descripcion
- ¿Hay alguna columna que no tenga sentido su contenido?
> La columna time no parece que tenga mucho sentido. Hay una mezcla entre horas y texto, y tampoco tiene mucha importancia
- ¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?
> El time que hemos comentado antes, debería estar agrupado por horas o incluso por mañana, tarde o noche

> La de species tiene también demasiadas categorías, habría que resumirlas en unas pocas especies, o tipo de tiburón. Algo más general

> La columna activity también tiene demasiadas categorías, habría que resumirlas en unas pocas